base/strings/basic.jl

```diff
@@ -565,25 +565,15 @@ isascii(c::Char) = bswap(reinterpret(UInt32, c)) < 0x80
 isascii(s::AbstractString) = all(isascii, s)
 isascii(c::AbstractChar) = UInt32(c) < 0x80
 
-## string map, filter, has ##
-function map(f, s::AbstractString)
-    out = IOBuffer(sizehint=sizeof(s))
-    for c in s
-        c′ = f(c)
-        isa(c′, AbstractChar) || throw(ArgumentError(
-            "map(f, s::AbstractString) requires f to return AbstractChar; try map(f, collect(s)) or a comprehension instead"))
-        write(out, c′::AbstractChar)
-    end
-    String(take!(out))
-end
+## string map, filter ##
 
-function map(f, s::String)
+function map(f, s::AbstractString)
     out = StringVector(max(4, sizeof(s)÷sizeof(codeunit(s))))
     index = UInt(1)
     for c in s
         c′ = f(c)
-        isa(c′, Char) || throw(ArgumentError(
-            "map(f, s::String) requires f to return Char; " *
+        isa(c′, AbstractChar) || throw(ArgumentError(
+            "map(f, s::AbstractString) requires f to return AbstractChar; " *
             "try map(f, collect(s)) or a comprehension instead"))
         index + 3 > length(out) && resize!(out, unsigned(2 * length(out)))
         index += __unsafe_string!(out, c′, index)
```

Comment:
`__unsafe_string!` only accepts a `Char`; should we call `convert`?

Reply:
I guess? I don't really know what the an AbstractChar has to support.

Reply:
`convert` seems like a good idea.   The worst that can happen is that it throws a `MethodError`, which should at least be fairly clear.